### PR TITLE
Decompose AvgPool3d Op to Conv3d for Support

### PR DESCRIPTION
**Pull request by kamalrajkannan78**
## Summary:

- This PR fixes [#713](https://github.com/tenstorrent/tt-forge-fe/issues/713)

- The core idea behind this transformation is to use Conv3d as a replacement for AvgPool3d by setting the convolution kernel's weights to a constant value. Specifically, the weight for each element in the kernel is set to:

```
weight_value = 1.0 / (kD * kH * kW)
weight = torch.ones(1, 1, kD, kH, kW) * weight_value
```

- This approach effectively turns the convolution into an average pooling operation, as the weights are identical, and their sum over the receptive field will produce the average. (Note : conv3d is not yet supported in MLIR and TTNN)

- eval,shape,decompose fucntions and sanity tests are added for conv3d, avgpool3d

## Logs

- [avgpool3d_test_case_results.log](https://github.com/user-attachments/files/17772062/nov15_avgpool3d_all_cases_f1.log)
- [conv3d_test_case_results.log](https://github.com/user-attachments/files/17772063/nov15_conv3d_all_cases_f1.log)

### CODE REVIEW

**Review by dgolubovicTT — APPROVED** [review]

**Comment by dgolubovicTT** [comment]
@kamalrajkannan78 @nvukobratTT Can you share why we decompose avgpool3d? Since tt-metal has avgpool2d, can we expect it to have avgpool3d support, or ask for it? Do we know if they had the need for avgpool3d themselves? 
This decomposition is logical but whenever I see some decomposition lately, I start thinking about how performant decomposed subgraph is compared to native op in ttnn. So if we do decompose, we are less overfitting to ttnn, and if we don't do it and require ttnn support to lower directly, we are maybe overfitting to ttnn. 

What is your take in this case @nvukobratTT

**Comment by nvukobratTT** [comment]
> @kamalrajkannan78 @nvukobratTT Can you share why we decompose avgpool3d? Since tt-metal has avgpool2d, can we expect it to have avgpool3d support, or ask for it? Do we know if they had the need for avgpool3d themselves? This decomposition is logical but whenever I see some decomposition lately, I start thinking about how performant decomposed subgraph is compared to native op in ttnn. So if we do decompose, we are less overfitting to ttnn, and if we don't do it and require ttnn support to lower directly, we are maybe overfitting to ttnn.
> 
> What is your take in this case @nvukobratTT

Few notes before more detailed CR:
1. I'm okay with doing these decompositions as TTNN lacks support for these ops. Mostly because it's questionable when (if) they are going to add those. That said, @kamalrajkannan78 can you open two issues on tt-metal repo with a op support request and like them here? Ideally, we should like to have those ops on TTNN as well. 
2. Can we decompose avg pool 3d using avg pool 2d? Current solution looks mathematically correct, however, it will impose confusions when when someone starts to debug what is happening here. From my side, it seems cleaner to decompose conv3d using conv2d, and doing similar for avg pool 3d using avg pool 2d. What are your thoughts guys?
3. For avg pool 3d we can use a similar decomposition approach:
    - Apply avg pool 2d over on each depth slice (keep intermediate results for each slice)
    - Stack those results along depth dim
    - Average those results across depth dim

**Comment by kamalrajkannan78** [comment]
> @kamalrajkannan78 @nvukobratTT Can you share why we decompose avgpool3d? Since tt-metal has avgpool2d, can we expect it to have avgpool3d support, or ask for it? Do we know if they had the need for avgpool3d themselves? This decomposition is logical but whenever I see some decomposition lately, I start thinking about how performant decomposed subgraph is compared to native op in ttnn. So if we do decompose, we are less overfitting to ttnn, and if we don't do it and require ttnn support to lower directly, we are maybe overfitting to ttnn.
> 
> What is your take in this case @nvukobratTT

thanks for the review @dgolubovicTT 

- Alexnet model faced `Encountered unsupported op node type: nn.avg_pool3d, on device: tt issue ` .  I looked into nn.avg_pool2d decomposition and followed similar way to decompose nn.avg_pool3d. 
- Inorder to support alexnet osmr model avgpool3d is needed & It will be helpful if we get a support from MLIR for this op
- If we decompose it with some other available ops , we may need to add support in those available ops to handle our 5d input tensor.

**Comment by kamalrajkannan78** [comment]
> > @kamalrajkannan78 @nvukobratTT Can you share why we decompose avgpool3d? Since tt-metal has avgpool2d, can we expect it to have avgpool3d support, or ask for it? Do we know if they had the need for avgpool3d themselves? This decomposition is logical but whenever I see some decomposition lately, I start thinking about how performant decomposed subgraph is compared to native op in ttnn. So if we do decompose, we are less overfitting to ttnn, and if we don't do it and require ttnn support to lower directly, we are maybe overfitting to ttnn.
> > What is your take in this case @nvukobratTT
> 
> Few notes before more detailed CR:
> 
> 1. I'm okay with doing these decompositions as TTNN lacks support for these ops. Mostly because it's questionable when (if) they are going to add those. That said, @kamalrajkannan78 can you open two issues on tt-metal repo with a op support request and like them here? Ideally, we should like to have those ops on TTNN as well.
> 2. Can we decompose avg pool 3d using avg pool 2d? Current solution looks mathematically correct, however, it will impose confusions when when someone starts to debug what is happening here. From my side, it seems cleaner to decompose conv3d using conv2d, and doing similar for avg pool 3d using avg pool 2d. What are your thoughts guys?
> 3. For avg pool 3d we can use a similar decomposition approach:
>    
>    * Apply avg pool 2d over on each depth slice (keep intermediate results for each slice)
>    * Stack those results along depth dim
>    * Average those results across depth dim


sure @nvukobratTT I will try to decompose avgpool3d in a suggested way!

**Comment by dgolubovicTT** [comment]
Decomposing avgpool3d to avgpool2d like @nvukobratTT layed out sounds great. Let's do it like that!

**Comment by kamalrajkannan78** [comment]
> > @kamalrajkannan78 @nvukobratTT Can you share why we decompose avgpool3d? Since tt-metal has avgpool2d, can we expect it to have avgpool3d support, or ask for it? Do we know if they had the need for avgpool3d themselves? This decomposition is logical but whenever I see some decomposition lately, I start thinking about how performant decomposed subgraph is compared to native op in ttnn. So if we do decompose, we are less overfitting to ttnn, and if we don't do it and require ttnn support to lower directly, we are maybe overfitting to ttnn.
> > What is your take in this case @nvukobratTT
> 
> Few notes before more detailed CR:
> 
> 1. I'm okay with doing these decompositions as TTNN lacks support for these ops. Mostly because it's questionable when (if) they are going to add those. That said, @kamalrajkannan78 can you open two issues on tt-metal repo with a op support request and like them here? Ideally, we should like to have those ops on TTNN as well.
> 2. Can we decompose avg pool 3d using avg pool 2d? Current solution looks mathematically correct, however, it will impose confusions when when someone starts to debug what is happening here. From my side, it seems cleaner to decompose conv3d using conv2d, and doing similar for avg pool 3d using avg pool 2d. What are your thoughts guys?
> 3. For avg pool 3d we can use a similar decomposition approach:
>    
>    * Apply avg pool 2d over on each depth slice (keep intermediate results for each slice)
>    * Stack those results along depth dim
>    * Average those results across depth dim

- https://github.com/tenstorrent/tt-metal/issues/15102
- https://github.com/tenstorrent/tt-metal/issues/15103

File: forge/forge/op/eval/forge/convolution.py [hunk 1]
```diff
@@ -372,3 +372,194 @@ def is_eltwise_unary(self) -> bool:
...
+            )
+
+            if not is_channel_last:
+                result = dc.op(TransposeTM.create(dim0=-1, dim1=-2), [result])
+                result = dc.op(TransposeTM.create(dim0=-2, dim1=-3), [result])
+                result = dc.op(TransposeTM.create(dim0=-3, dim1=-4), [result])
+            dc.fuse(result)
+
+    def backward(self, ac, operand, inputs, output, grad):
+        pass
```
Comment (nvukobratTT) [inline forge/forge/op/eval/forge/convolution.py:547]: Can we raise exception here that it's not supported?

File: forge/test/mlir/test_ops.py [hunk 2]
```diff
@@ -14,6 +14,89 @@
...
+    fw_out = framework_model(*inputs)
+    compiled_model = forge.compile(framework_model, sample_inputs=inputs)
+    co_out = compiled_model(*inputs)
+
+    co_out = [co.to("cpu") for co in co_out]
+    fw_out = [fw_out] if isinstance(fw_out, torch.Tensor) else fw_out
+    assert all([compare_with_golden_pcc(golden=fo, calculated=co, pcc=0.99) for fo, co in zip(fw_out, co_out)])
+
+
+@pytest.mark.parametrize(
```
Comment (nvukobratTT) [inline forge/test/mlir/test_ops.py:58]: It'll be good to test out channel last for conv3d as well

File: forge/test/mlir/test_ops.py [hunk 3]
```diff
@@ -14,6 +14,89 @@
...
+        ((1, 1, 100, 54, 54), (5, 1, 1), (1, 1, 1), (0, 0, 0), (1, 1, 1), 1),
+        ((1, 1, 32, 32, 32), (3, 3, 3), (1, 1, 1), (1, 1, 1), (1, 1, 1), 2),
+        ((1, 1, 50, 28, 28), (3, 3, 3), (2, 2, 2), (1, 1, 1), (1, 1, 1), 3),
+        ((1, 1, 32, 32, 32), (3, 3, 3), (1, 1, 1), (1, 1, 1), (2, 2, 2), 4),
+        ((1, 1, 20, 20, 20), (3, 3, 3), (1, 1, 1), (2, 2, 2), (1, 1, 1), 5),
+        ((1, 1, 8, 8, 8), (1, 1, 1), (1, 1, 1), (0, 0, 0), (1, 1, 1), 7),
+        ((1, 1, 16, 16, 16), (5, 5, 5), (1, 1, 1), (2, 2, 2), (1, 1, 1), 8),
+    ],
+)
+@pytest.mark.xfail(reason="Found Unsupported operations while lowering from TTForge to TTIR in forward graph : conv3d")
```
Comment (nvukobratTT) [inline forge/test/mlir/test_ops.py:70]: Shouldn't this be decomposed into multiple conv2d ops?

**Comment by kamalrajkannan78** [comment]
> Thanks for the issues @kamalrajkannan78 !
> 
> FYI, I updated labels nad added project information to reflect more precisely what our request is :))

Thanks for adding those details @nvukobratTT

File: forge/test/mlir/test_ops.py (hunk 3 above)
Comment (kamalrajkannan78) [reply forge/test/mlir/test_ops.py:70]: decomposition of conv3d to conv2d is not yet done in forge. From bringup perspective, alexnet model has avgpool3d and its decomposition is needed now and currently I am working on decomposing it with avgpool 2d. if conv3d decomposition is also needed, I will comeback to this after finishing that.  is it okay? @nvukobratTT

**Comment by kamalrajkannan78** [comment]
> > @kamalrajkannan78 @nvukobratTT Can you share why we decompose avgpool3d? Since tt-metal has avgpool2d, can we expect it to have avgpool3d support, or ask for it? Do we know if they had the need for avgpool3d themselves? This decomposition is logical but whenever I see some decomposition lately, I start thinking about how performant decomposed subgraph is compared to native op in ttnn. So if we do decompose, we are less overfitting to ttnn, and if we don't do it and require ttnn support to lower directly, we are maybe overfitting to ttnn.
> > What is your take in this case @nvukobratTT
> 
> Few notes before more detailed CR:
> 
> 1. I'm okay with doing these decompositions as TTNN lacks support for these ops. Mostly because it's questionable when (if) they are going to add those. That said, @kamalrajkannan78 can you open two issues on tt-metal repo with a op support request and like them here? Ideally, we should like to have those ops on TTNN as well.
> 2. Can we decompose avg pool 3d using avg pool 2d? Current solution looks mathematically correct, however, it will impose confusions when when someone starts to debug what is happening here. From my side, it seems cleaner to decompose conv3d using conv2d, and doing similar for avg pool 3d using avg pool 2d. What are your thoughts guys?
> 3. For avg pool 3d we can use a similar decomposition approach:
>    
>    * Apply avg pool 2d over on each depth slice (keep intermediate results for each slice)
>    * Stack those results along depth dim
>    * Average those results across depth dim

https://github.com/tenstorrent/tt-forge-fe/pull/741
